Setup SNP on the RHEL host

LakshmiSaiHarika opened the pull request:
Updated snp.sh to install dependencies, build SNP latest kernel package and set the default grub menu to the SNP latest kernel version in RHEL (Tested this on RHEL 9.4)

LakshmiSaiHarika committed:
Update snp docs for RHEL

Added RHEL 9.4 OS as one of the tested OS distribution for sev-utils

Signed-off-by: Harika Nittala <lnittala@amd.com>

ryansavino commented on docs/snp.md:
````diff
@@ -51,6 +52,11 @@ Download the script and add the execute permission:
 wget https://github.com/amd/sev-utils/raw/main/tools/snp.sh
 chmod +x snp.sh
 ```
+For RHEL OS distribution, export the following environment variables for using your RHEL subscription manager credentials for installing the package dependencies:
````
This shouldn't be necessary. If you're installing packages on the host, let it error out if subscription manager isn't enabled.

LakshmiSaiHarika replied:
Done, I removed this portion

ryansavino replied:
I still see the variables and method to register the subscription manager.

LakshmiSaiHarika replied:
Ok, Now I completely removed RH subscription manager part from the script

ryansavino commented on tools/snp.sh:
```diff
@@ -90,6 +90,10 @@ QEMU_CMDLINE_FILE="${QEMU_CMDLINE:-${LAUNCH_WORKING_DIR}/qemu.cmdline}"
 IMAGE="${IMAGE:-${LAUNCH_WORKING_DIR}/${GUEST_NAME}.img}"
 GENERATED_INITRD_BIN="${SETUP_WORKING_DIR}/initrd.img"
 
+# Redhat Sensitive Credentials -- User must specify, no default value set
```
See docs comment. Shouldn't be needed.

LakshmiSaiHarika replied:
Done, I removed this line

ryansavino commented on tools/snp.sh:
```diff
@@ -214,7 +215,41 @@ install_sev_snp_measure() {
 }
 
-install_dependencies() {
+check_if_redhat_credentials_set(){
```
Instead, maybe have a function to check if the subscription manager is registered and working before proceeding.

LakshmiSaiHarika replied:
Created check_rhel_subscription_mgr_status() to check the status, if status is in unregister keyterms, then register RHEL subscription with the given RHEL env variables

LakshmiSaiHarika replied:
I removed this function as this script automatically errors out if RH subscription manager is unset

ryansavino commented on tools/snp.sh:
```diff
@@ -237,7 +286,9 @@ install_dependencies() {
-  #sudo apt install -y nasm
+
+  # Remove any default nasm package
+  sudo apt purge nasm
```
why do we need this?

LakshmiSaiHarika replied:
Removed this line as this works on fresh os image

ryansavino commented on tools/snp.sh:
```diff
@@ -268,10 +319,110 @@ install_dependencies() {
   sudo apt install -y python3-pip
+}
+
+check_if_redhat_credentials_set(){
```
Do you have this method twice?

LakshmiSaiHarika replied:
Done.. I cleaned this portion

ryansavino commented on tools/snp.sh:
```diff
@@ -268,10 +319,110 @@ install_dependencies() {
+
+  # Build dependencies
+   # Group install on RH provisioned RH Host
+  sudo dnf groupinstall -y "Headless Management" "Development Tools" "Container Management" "RPM Development Tools"
```
This is prob a lot of packages. Are you sure we need all of them? I would much rather have us install only required packages by name.

LakshmiSaiHarika replied:
Done, I removed groupinstall

ryansavino commented on tools/snp.sh:
```diff
@@ -268,10 +319,110 @@ install_dependencies() {
+  sudo dnf install -y iasl
+
+  # Remove any default nasm package
+  sudo dnf remove nasm
```
Why?

LakshmiSaiHarika replied:
Same, I removed this dnf remove nasm line

ryansavino commented on tools/snp.sh:
```diff
@@ -268,10 +319,110 @@ install_dependencies() {
+  # sev-snp-measure
+  sudo dnf install -y python3-pip
+
+  sudo dnf install -y curl
```
put this above with wget - only if you need it. Do you use it somewhere?

LakshmiSaiHarika replied:
Yes  curl is used for rust installation process in RHEL

ryansavino commented on tools/snp.sh:
```diff
@@ -268,10 +319,110 @@ install_dependencies() {
+  sudo dnf install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-9.noarch.rpm
+  sudo dnf install -y genisoimage
+
+  sudo dnf install -y curl
```
again?

LakshmiSaiHarika replied:
Removed duplicate curl
Also, I removed genisoimage for the moment as this package is used in the launch SNP guest process, which I will address in future PR

ryansavino commented on tools/snp.sh:
```diff
@@ -268,10 +319,110 @@ install_dependencies() {
+  sudo dnf install -y curl
+
+  # Installing additional dependencies on RHEL Host provisioned from satellite
+  sudo dnf install -y bc wget
```
wget again? satellite? If you don't need it, get rid of this.

Install the bare minimum required packages. Go through and install things one by one again just to see what's really needed.

LakshmiSaiHarika replied:
we require wget utility in RHEL to install nasm from the source and download the base guest image

ryansavino commented on tools/snp.sh:
```diff
@@ -268,10 +319,110 @@ install_dependencies() {
+  sudo dnf install -y curl
+
+  # Handles grub menu config
+  sudo dnf info -y grubby
```
info?

LakshmiSaiHarika replied:
Removed grubby package 
Modified RHEL grub default via rhel_set_grub_default_snp() to edit grub config file and reload grub config for UEFI/BIOS based machine

ryansavino commented on tools/snp.sh:
```diff
@@ -268,10 +319,110 @@ install_dependencies() {
+    # genisoimage utility supported until RedHat version 8
+    # For genisoimage utility support in RedHat 9, using epel
+  sudo dnf install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-9.noarch.rpm
+  sudo dnf install -y genisoimage
```
we need to avoid adding a repo (esp epel) unless completely necessary. Let's try to resolve this to an existing package. If not, I'll help you find a solution.

LakshmiSaiHarika replied:
Right now removing genisoimage from this PR (RHEL setup host PR) as this package is used in the launch-guest process..

I will update further dependencies for the SNP guest launch in the future PR

ryansavino commented on tools/snp.sh:
```diff
@@ -268,10 +319,110 @@ install_dependencies() {
+  sudo dnf install -y bc wget
 }
 
+
```
remove

LakshmiSaiHarika replied:
Done.. I cleaned this up

ramagali24 replied:
Since we have an install.sh in AMDSEV, can we move the installation of tools like Rust, NASM, and other dependencies from snp.sh into this file? As we continue to support more distributions, snp.sh is becoming more complicated. Please check this: https://github.com/ramagali24/AMDSEV-SVSM/blob/svsm-latest-fixes/install.sh, which includes Fedora and openSUSE. @ryansavino, I wanted to discuss this with you. What do you suggest?

ryansavino commented on tools/snp.sh:
```diff
@@ -1113,6 +1264,7 @@ main() {
       ;;
 
     setup-host)
+      identify_linux_distribution_type
```
Maybe we run this, only when needed?

LakshmiSaiHarika replied:
Done.. I updated this

LakshmiSaiHarika replied:
I updated identify_linux_distribution_type to get_linux_distro function

ryansavino commented on tools/snp.sh:
```diff
@@ -95,8 +95,8 @@ RHEL_SUBS_MGR_USER="${RHEL_SUBS_MGR_USER}"
-AMDSEV_URL="https://github.com/ryansavino/AMDSEV.git"
-AMDSEV_DEFAULT_BRANCH="snp-latest-fixes"
+AMDSEV_URL="https://github.com/LakshmiSaiHarika/AMDSEV.git"
+AMDSEV_DEFAULT_BRANCH="rhel-snp-latest-fixes"
```
what changes here? If not too involved maybe we can merge into my existing fork.

LakshmiSaiHarika replied:
AMDESE changes: In your fork, I added few fixes fir RHEL in build and install script

I raised a PR (below link) to merge in your AMDESE fork:
https://github.com/ryansavino/amdese-amdsev/pull/1

ryansavino commented on tools/snp.sh:
```diff
@@ -465,6 +465,28 @@ set_grub_default_snp() {
   sudo update-grub
 }
 
+grubby_to_set_grub_default_snp(){
```
grubby_set...

LakshmiSaiHarika replied:
Removed grubby from this script

Modified RHEL grub configuration and reloaded settings in "rhel_set_grub_default_snp"

LakshmiSaiHarika replied:
Commit Cleanup: I merged separate commits for setting grub menu into a single commit

ryansavino commented on tools/snp.sh:
```diff
@@ -214,7 +214,27 @@ install_sev_snp_measure() {
+    rhel)
+    LINUX_TYPE='rhel'
+
+    if [[ "$UPM" = false ]]; then
```
This check doesn't belong here. You can run a similar failure case outside this function somewhere if you want to make sure rhel doesn't proceed.

Missing break for `rhel` case.

indent under each category.

LakshmiSaiHarika replied:
I removed this non- UPM conditional check from this RHEL case

LakshmiSaiHarika committed:
snp.sh: Add Redhat dependencies installation prior to the SNP kernel build

Added RHEL dependencies to build the SNP kernel rpm package for the host and guest

Signed-off-by: Harika Nittala <lnittala@amd.com>

LakshmiSaiHarika committed:
snp.sh: Remove nasm package uninstallation step

Removed ubuntu nasm package uninstallation step from the nasm source installation process

Signed-off-by: Harika Nittala <lnittala@amd.com>

LakshmiSaiHarika committed:
snp.sh: Set GRUB default menu to built SNP kernel

Generalized set_default_grub_menu across various OS distribution
Renamed ubuntu_set_default_grub_menu to set specific default ubuntu grub menu to the built SNP kernel
Added a function to set the RHEL default grub menu to the built SNP kernel

Signed-off-by: Harika Nittala <lnittala@amd.com>

LakshmiSaiHarika commented on tools/snp.sh:
```diff
@@ -91,8 +91,8 @@ IMAGE="${IMAGE:-${LAUNCH_WORKING_DIR}/${GUEST_NAME}.img}"
-AMDSEV_URL="https://github.com/ryansavino/AMDSEV.git"
-AMDSEV_DEFAULT_BRANCH="snp-latest-fixes"
+AMDSEV_URL="https://github.com/LakshmiSaiHarika/AMDSEV.git"
+AMDSEV_DEFAULT_BRANCH="rhel-setup-host"
```
@ryansavino Please review and merge my PR for adding minor RHEL fixes into your AMDESE fork
PR link: [here](https://github.com/ryansavino/amdese-amdsev/pull/1)

LakshmiSaiHarika replied:
I cleaned up my AMDESE RHEL fixes ... please review this PR and let me know if any changes are needed

LakshmiSaiHarika commented on tools/snp.sh:
```diff
@@ -173,9 +173,6 @@ install_nasm_from_source() {
   fi
 
-  # Remove package manager nasm
-  sudo apt purge nasm
```
This part was already present in the initial snp.sh commit..so, I removed this

LakshmiSaiHarika commented on tools/snp.sh:
```diff
@@ -274,7 +263,86 @@ install_dependencies() {
-  
+}
+
+rhel_install_dependencies() {
```
Updated RHEL dependencies to minimally install required packages to setup SNP on RHEL host

LakshmiSaiHarika replied:
Also, I removed sudo subscription-manager commands completely from the script

LakshmiSaiHarika committed:
snp.sh: Update ubuntu dependency installation structure

Renamed old install_dependencies feature to ubuntu_install_dependencies
Updated install_dependencies to perform installation of dependencies based on identified linux distribution and their package manager command

Signed-off-by: Harika Nittala <lnittala@amd.com>

DGonzalezVillal reviewed (approved):
This looks good to me
